crowbar_framework/app/models/dns_service.rb

```diff
@@ -188,4 +188,63 @@ def save_config_to_databag(old_role, role, server_nodes = nil)
     instance = Crowbar::DataBagConfig.instance_from_role(old_role, role)
     Crowbar::DataBagConfig.save("core", instance, @bc_name, config)
   end
+
+  # try to know if we can skip a node from running chef-client
+  def skip_unchanged_node?(node, old_role, new_role)
+    @logger.debug("DNS skip_batch_for_node? entering: #{node}")
+
+    # if old_role is nil, then we are applying the barclamp for the first time
+    if old_role.nil?
+      @logger.debug("DNS skip_batch_for_node?: not skipping #{node} (new role)")
+      return false
+    end
+
+    # if attributes have changed, we need to run
+    return false if node_changed_attributes?(node, old_role, new_role)
+
+    # if the node changed roles, then we need to apply
+    return false if node_changed_roles?(node, old_role, new_role)
+
```

Comment:
In the case of dns, I would try to optimize things further and skip client-only nodes unless some attributes are changed (nameservers, additional_search_domains). I think all other attributes only matter for dns-server. This makes it possible to change the list of records without running chef of client nodes.